src/mpid/ch4/shm/posix/eager/iqueue/iqueue_recv.h

```diff
@@ -142,16 +142,27 @@ MPL_STATIC_INLINE_PREFIX void
 MPIDI_POSIX_eager_recv_commit(MPIDI_POSIX_eager_recv_transaction_t * transaction)
 {
     MPIDI_POSIX_eager_iqueue_cell_t *cell;
+    MPIDI_POSIX_eager_iqueue_transport_t *transport;
 
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_IQUEUE_RECV_COMMIT);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_IQUEUE_RECV_COMMIT);
 
+    /* Get the transport from the global variables */
+    transport = MPIDI_POSIX_eager_iqueue_get_transport();
+
     cell = (MPIDI_POSIX_eager_iqueue_cell_t *) transaction->transport.iqueue.pointer_to_cell;
-    MPIR_Assert(cell != NULL);
+    MPIR_Assert(cell != NULL && cell->type != MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_NULL);
     cell->next = NULL;
     cell->prev = 0;
-    MPL_atomic_compiler_barrier();
-    cell->type = MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_NULL;
+
+    if (transport->last_read_cell) {
+        /* Free previously last read cell */
+        MPIR_Assert(transport->last_read_cell->type != MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_NULL);
+        transport->last_read_cell->type = MPIDI_POSIX_EAGER_IQUEUE_CELL_TYPE_NULL;
+    }
+
+    /* Save a new last read cell */
+    transport->last_read_cell = cell;
```

Comment:
Just trying do understand this optimization. Shouldn't `last_read_cell` be in shared memory and be accessed atomically?

Reply:
No, it's just trying to capture the last cell used by the local process because it's most likely to be cache hot.

Reply:
I may be remembering things imprecisely so correct me if I am wrong. Let's say process 0 is sending data to process 1. Process 0 gets a cell from its own list of cells, fills it with data and sends it to process 1 by adding it to process 1's terminal. Process 1 dequeues the cell from its terminal, processes it and then releases it so that it can be reused by process 0. In this case this code is also saving a cell reference locally in the transport of process 1 (`last_read_cell`). If process 1 immediately after the receive operation wants to send data to process 2 it tries to get `last_read_cell` from the transport, which however belongs to process 0. Is this correct? Here comes my confusion, is this supposed to happen?

Reply:
Ah. I had it backward. This is the last cell being read so the sender wants to grab it because its more likely to be empty, not because it's more likely to be in cache.

Anyway, it doesn't need to be in cache because the cell it is pointing to is the last one that was read by the local process, meaning it's already a shared memory pointer. The iqueue cells are all in shared memory and the one last read was last in use by the local process, so it's most likely to be empty. "Empty lists" aren't specific to a process, but are shared by all processes. The "terminals" where the full cells go are specific to a receiving process. I think some of the comments here are incorrect. I wrote most of the comments, but I didn't write most of the code and I think as I've learned more about the code, I'm seeing where I was wrong when I wrote the original comments.

As for why it's not protected by a lock, that's less clear to me. This particular assignment is safe (I think) on x86, but the check before it probably isn't. Nor is the one in https://github.com/pmodels/mpich/pull/4368/files#diff-6e2727273115a31a36a396d997215256R24-R31. It might not be safe on other architectures. I was recently having this conversation with @pavanbalaji. Maybe @ddurnov or @SergeyGubanov can help out here.

Reply:
> If process 1 immediately after the receive operation wants to send data to process 2 it tries to get `last_read_cell` from the transport, which however belongs to process 0. Is this correct? 

Yes, it is correct.

To avoid hangs, the number of cells per rank must be greater than the number of ranks.

Reply:
Just to further clarify my understanding. This is not the same model used by CH3 in which free lists, and cells within them, are used exclusively by the owning process. I am also not sure the assumption on the number of cells greater than the number local ranks is correct as in CH4's code the number of cells allocated does not depend on the number of local processes currently.

Reply:
That's correct.